src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -0,0 +1,197 @@
+        }
+        StringBuilder pattern = new StringBuilder();
+        int count = 0;
+        for (Map.Entry<String,String> ent : variables.entrySet()) {
+            String val = ent.getValue();
+            if (val == null || val.isEmpty() || safeEnvVariables.contains(ent.getKey())) {  // Skip values that are safe
+                continue;
+            }
+            if (count > 0) {
+                pattern.append('|');
+            }
+            pattern.append(Pattern.quote(val));
+            count++;
+        }
+        return (count > 0)
+                ? !Pattern.compile(pattern.toString()).matcher(input).find()
+                : true;
+    }
+
+    private static final HashSet<String> SAFE_ENVIRONMENT_VARIABLES = new HashSet<String>(Arrays.asList(
+            // Pipeline/Jenkins variables in normal builds
+            "BRANCH_NAME",
+            "BUILD_DISPLAY_NAME",
+            "BUILD_ID",
+            "BUILD_NUMBER",
+            "BUILD_TAG",
+            "BUILD_URL",
+            "CHANGE_AUTHOR",
+            "CHANGE_AUTHOR_DISPLAY_NAME",
+            "CHANGE_AUTHOR_EMAIL",
+            "CHANGE_ID",
+            "CHANGE_TARGET",
+            "CHANGE_TITLE",
+            "CHANGE_URL",
+            "EXECUTOR_NUMBER",
+            "HUDSON_COOKIE",
+            "HUDSON_HOME",
+            "HUDSON_SERVER_COOKIE",
+            "HUDSON_URL",
+            "JENKINS_HOME",
+            "JENKINS_SERVER_COOKIE",
+            "JENKINS_URL",
+            "JOB_BASE_NAME",
+            "JOB_NAME",
+            "JOB_URL",
+            "NODE_LABELS",
+            "NODE_NAME",
+            "WORKSPACE",
+
+            // Normal system variables for posix environments
+            "HOME",
+            "LANG",
+            "LOGNAME",
+            "MAIL",
+            "NLSPATH",
+            "PATH",
+            "PWD",
+            "SHELL",
+            "SHLVL",
+            "TERM",
+            "USER",
+            "XFILESEARCHPATH"
+    ));
+
+    /**
+     * Does first-level sanitization of a step's params, returning the Describable arguments map
+     * @param stepArguments Arguments provided when creating the step
+     * @param variables Environment variables to remove
+     * @return
+     */
+    @Nonnull
+    public static Map<String,Object> sanitizedStepArguments(@Nonnull Map<String, Object> stepArguments, @CheckForNull EnvVars variables) {
+        HashMap<String, Object> output = new HashMap<String, Object>();
+
+        if (variables == null || variables.size() == 0) {
+            // No need to sanitize against environment variables
+            for (Map.Entry<String,Object> entry : stepArguments.entrySet()) {
+                if (entry.getValue() instanceof CharSequence && ((CharSequence)(entry.getValue())).length() > ArgumentsAction.MAX_STRING_LENGTH) {
+                    output.put(entry.getKey(), OVERSIZE_VALUE);
+                } else {
+                    output.put(entry.getKey(), entry.getValue());
```

Comment:
And as a defense in case something like that was forgotten, you might need to do something like https://github.com/jenkinsci/workflow-api-plugin/pull/18.